src/wallet/spend.cpp

```diff
@@ -106,7 +106,13 @@ void CoinsResult::Erase(const std::unordered_set<COutPoint, SaltedOutpointHasher
 {
     for (auto& [type, vec] : coins) {
         auto remove_it = std::remove_if(vec.begin(), vec.end(), [&](const COutput& coin) {
-            return coins_to_remove.count(coin.outpoint) == 1;
+            // remove it if it's on the set
+            if (coins_to_remove.count(coin.outpoint) == 0) return false;
+
+            // update cached amounts
+            total_amount -= coin.txout.nValue;
+            if (coin.HasEffectiveValue()) total_effective_amount = *total_effective_amount - coin.GetEffectiveValue();
```

Comment:
Are we sure that `total_effective_value` is always set here?

Maybe `assume(total_effective_amount.has_value()` here?

Reply:
I had this question myself. I figured if we remove the coin with effective value, then the coin was added before, hence we have `total_effective_value` defined. Please correct me if my reasoning is flawed.

Reply:
yeah. As we only use `CoinsResult::Add` to append new coins: if we have the coin inside `CoinsResult`, and the coin has an effective value, then we must have `total_effective_amount`.